plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,314 @@
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (transitionsToIdleUponUserInteraction) {
+        currentCancelable?.cancel()
+        currentCancelable = null
+        updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION)
+      }
+    }
+
+    override fun onMove(detector: MoveGestureDetector): Boolean {
+      // no-ops
+      return false
+    }
+
+    override fun onMoveEnd(detector: MoveGestureDetector) {
+      // no-ops
+    }
+  }
+
+  /**
+   * Returns list of registered states.
+   */
+  override val states: Set<ViewportState>
+    get() = internalStates.toSet()
+
+  /**
+   * Returns current [ViewportStatus].
+   *
+   * If current status is IDLE, returns ViewportStatus.State(null).
+   * @see addStatusObserver
+   */
+  override var status: ViewportStatus = ViewportStatus.State(null)
+    private set
+
+  private fun updateStatus(targetStatus: ViewportStatus, reason: String) {
+    val previousStatus = status
+    status = targetStatus
+    notifyStatusChanged(previousStatus, targetStatus, reason)
+  }
+
+  /**
+   * Executes a transition to requested state.
+   *
+   * When started, goes to [ViewportTransition]
+   * and to the final [ViewportState] when ended.
+   *
+   * If transition is canceled, state goes to IDLE.
+   *
+   * @param targetState The target [ViewportState] to transition to.
+   * @param completionListener The listener to observe the completion state.
+   */
+  override fun transitionTo(targetState: ViewportState, completionListener: CompletionListener?) {
+    if (status.getTargetState() === targetState) {
+      completionListener?.onComplete(false)
+      return
+    }
+    val fromState = status.getTargetState()
+    currentCancelable?.cancel()
+    currentCancelable = null
+    if (!states.contains(targetState)) {
+      throw IllegalStateException("Target ViewportState is not added.")
+    }
+
+    // get the transition (or default) for the from and to state
+    val transition = getTransition(fromState, targetState) ?: defaultTransition
+
+    // run the transition
+    var completionBlockInvoked = false
+    val transitionCancelable = transition.run(fromState, targetState) { isCanceled ->
+      completionBlockInvoked = true
+      if (!isCanceled) {
+        // transfer camera updating responsibility to targetState
+        currentCancelable = targetState.startUpdatingCamera()
+        updateStatus(
+          ViewportStatus.State(targetState),
+          VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+        )
+      }
+      completionListener?.onComplete(isCanceled)
+    }
+    // since it's possible that a transition might invoke its
+    // completion block synchronously, we'll only store the
+    // transition cancelable if the transition is not complete
+    // so that we don't clobber the targetState cancelable.
+    if (!completionBlockInvoked) {
+      currentCancelable = transitionCancelable
+      updateStatus(
```

Comment:
If it's possible for the transition to run async - is the thread always the same?

Reply:
Added Synchronized annotation to the status and currentCancelable getter/setter functions.